velox/exec/RowNumber.cpp

```diff
@@ -182,4 +356,152 @@ int64_t RowNumber::numRows(char* partition) {
 }
+
+void RowNumber::reclaim(
+    uint64_t /*targetBytes*/,
+    memory::MemoryReclaimer::Stats& stats) {
+  if (table_ == nullptr || table_->numDistinct() == 0) {
+    // Nothing to spill.
+    return;
+  }
+
+  if (hashTableSpiller_) {
+    // Already spilled.
+    return;
+  }
+
+  if (nonReclaimableSection_) {
+    ++stats.numNonReclaimableAttempts;
+    return;
+  }
+
+  spill();
+}
+
+void RowNumber::setupHashTableSpiller() {
+  // TODO Replace joinPartitionBits and Spiller::Type::kHashJoinBuild.
+
+  const auto& spillConfig = spillConfig_.value();
+  HashBitRange hashBits(
+      spillConfig.startPartitionBit,
+      spillConfig.startPartitionBit + spillConfig.joinPartitionBits);
+
+  auto columnTypes = table_->rows()->columnTypes();
+  auto tableType = ROW(std::move(columnTypes));
+
+  hashTableSpiller_ = std::make_unique<Spiller>(
+      Spiller::Type::kHashJoinBuild,
+      table_->rows(),
+      [&](folly::Range<char**> rows) { table_->rows()->eraseRows(rows); },
+      tableType,
+      std::move(hashBits),
+      tableType->size() - 1,
+      std::vector<CompareFlags>(),
+      spillConfig.filePath,
+      spillConfig.maxFileSize,
+      spillConfig.writeBufferSize,
+      spillConfig.minSpillRunSize,
+      spillConfig.compressionKind,
+      memory::spillMemoryPool(),
+      spillConfig.executor);
+}
+
+void RowNumber::setupInputSpiller() {
+  const auto& spillConfig = spillConfig_.value();
+  const auto hashBits = hashTableSpiller_->hashBits();
+
+  // TODO Replace Spiller::Type::kHashJoinProbe.
+  inputSpiller_ = std::make_unique<Spiller>(
+      Spiller::Type::kHashJoinProbe,
+      inputType_,
+      std::move(hashBits),
+      spillConfig.filePath,
+      spillConfig.maxFileSize,
+      spillConfig.writeBufferSize,
+      spillConfig.minSpillRunSize,
+      spillConfig.compressionKind,
+      memory::spillMemoryPool(),
+      spillConfig.executor);
+
+  const auto& hashers = table_->hashers();
+
+  std::vector<column_index_t> keyChannels;
+  keyChannels.reserve(hashers.size());
+  for (const auto& hasher : hashers) {
+    keyChannels.push_back(hasher->channel());
+  }
+
+  spillHashFunction_ = std::make_unique<HashPartitionFunction>(
+      inputSpiller_->hashBits(), inputType_, keyChannels);
+}
+
+void RowNumber::spill() {
+  VELOX_CHECK(spillEnabled());
+  VELOX_CHECK_NULL(hashTableSpiller_);
+  VELOX_CHECK_NULL(inputSpiller_);
+
+  setupHashTableSpiller();
+  setupInputSpiller();
+
+  std::vector<Spiller::SpillableStats> spillableStats(
+      hashTableSpiller_->hashBits().numPartitions());
+  hashTableSpiller_->fillSpillRuns(spillableStats);
+  hashTableSpiller_->spill();
+  hashTableSpiller_->finishSpill(spillHashTablePartitionSet_);
+
+  table_->clear();
+  pool()->release();
+
+  inputSpiller_->setPartitionsSpilled(
+      hashTableSpiller_->state().spilledPartitionSet());
+
+  if (input_ != nullptr) {
+    spillInput(input_, memory::spillMemoryPool());
```

Comment:
That's even better to spill input. Nice! I thought to catch the output produced by the remaining input.